lib/luahid.c

```diff
@@ -0,0 +1,96 @@
+/*
+* SPDX-FileCopyrightText: (c) 2025 Jieming Zhou <qrsikno@gmail.com>
+* SPDX-License-Identifier: MIT OR GPL-2.0-only
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/hid.h>
+#include <lunatik.h>
+
+typedef struct luahid_s {
+	lunatik_object_t *runtime;
+	struct hid_driver driver;
+	bool registered;
+} luahid_t;
+
+static const struct hid_device_id luahid_table[] = {
+	{HID_DEVICE(HID_BUS_ANY, HID_GROUP_ANY, HID_ANY_ID, HID_ANY_ID)},
+	{ }
+};
+MODULE_DEVICE_TABLE(hid, luahid_table);
+
+static void luahid_release(void *private)
+{
+	luahid_t *hid = (luahid_t *)private;
+	if (hid->registered)
+		hid_unregister_driver(&hid->driver);
+	if (hid->runtime != NULL)
+		lunatik_putobject(hid->runtime);
+	if (hid->driver.name != NULL)
+		lunatik_free(hid->driver.name);
+}
+
+static int luahid_register(lua_State *L);
+
+static const luaL_Reg luahid_lib[] = {
+	{"register", luahid_register},
+	{NULL, NULL}
+};
+
+static const luaL_Reg luahid_mt[] = {
+	{"__gc", lunatik_deleteobject},
+	{NULL, NULL}
+};
+
+static const lunatik_class_t luahid_class = {
+	.name = "hid",
+	.methods = luahid_mt,
+	.release = luahid_release,
+	.sleep = true,
+};
+
+static int luahid_register(lua_State *L)
+{
+	luaL_checktype(L, 1, LUA_TTABLE);
+
+	lunatik_object_t *object = lunatik_newobject(L, &luahid_class, sizeof(luahid_t));
+	luahid_t *hid = (luahid_t *)object->private;
+	hid->registered = false;
+	hid->runtime = NULL;
+
+	struct hid_driver *user_driver = &(hid->driver);
+	user_driver->name = NULL;
+
+	user_driver->name = lunatik_checkalloc(L, NAME_MAX);
+	lunatik_setstring(L, 1, user_driver, name, NAME_MAX);
+	user_driver->id_table = luahid_table;
+
+	lunatik_setruntime(L, hid, hid);
+	lunatik_getobject(hid->runtime);
+
+	if (__hid_register_driver(user_driver, THIS_MODULE, KBUILD_MODNAME) != 0)
+		luaL_error(L, "failed to register hid driver: %s", user_driver->name);
+	else
+		hid->registered = true;
```

Comment:
I don't think we need an extra flag for this.. you can just move lunatik_setstring() to here, right?